File: app/src/main/java/jp/juggler/subwaytooter/ActCallback.java
package jp.juggler.subwaytooter;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.FileProvider;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;

import org.apache.commons.io.IOUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicReference;

import jp.juggler.subwaytooter.util.LogCategory;
import jp.juggler.subwaytooter.util.Utils;

public class ActCallback extends AppCompatActivity {
	private static final LogCategory log = new LogCategory( "ActCallback" );
	
	public static final String ACTION_NOTIFICATION_CLICK = "notification_click";
	
	static final AtomicReference< Uri > last_uri = new AtomicReference<>( null );
	static final AtomicReference< Intent > sent_intent = new AtomicReference<>( null );
	
	@Override protected void onCreate( @Nullable Bundle savedInstanceState ){
		super.onCreate( savedInstanceState );
		Intent intent = getIntent();
		if( intent != null ){
			String action = intent.getAction();
			String type = intent.getType();
			if( Intent.ACTION_SEND.equals( action )
				|| Intent.ACTION_SEND_MULTIPLE.equals( action )
				|| ( type != null && type.startsWith( "image/" ) && Intent.ACTION_VIEW.equals( action ) )
				){
				
				// Google Photo などから送られるIntentに含まれるuriの有効期間はActivityが閉じられるまで
				// http://qiita.com/pside/items/a821e2fe9ae6b7c1a98c
				
				// 有効期間を延長する
				intent = remake( intent );
				if( intent != null ){
					sent_intent.set( intent );
				}
			}else{
				Uri uri = intent.getData();
				if( uri != null ){
					last_uri.set( uri );
				}
			}
		}
		// どうであれメイン画面に戻る
		intent = new Intent( this, ActMain.class );
		intent.addFlags( Intent.FLAG_ACTIVITY_LAUNCHED_FROM_HISTORY | Intent.FLAG_ACTIVITY_NEW_TASK );
		startActivity( intent );
		finish();
	}
	
	@Nullable Intent remake( @NonNull Intent src ){
		
		sweepOldCache();
		
		try{
			final String action = src.getAction();
			final String type = src.getType();
			
			if( type != null && (type.startsWith( "image/" ) || type.startsWith( "video/" )) ){
				if( Intent.ACTION_VIEW.equals( action ) ){
					Uri uri = src.getData();
					if( uri == null ) return null;
					try{
						uri = saveToCache( uri );
						Intent dst = new Intent( action );
						dst.setDataAndType( uri, type );
						return dst;
					}catch( Throwable ex ){
						log.trace( ex );
					}
				}else if( Intent.ACTION_SEND.equals( action ) ){
					Uri uri = src.getParcelableExtra( Intent.EXTRA_STREAM );
					if( uri == null ){
						// text/plain
						return src;
					}else{
						try{
							uri = saveToCache( uri );
							
							Intent dst = new Intent( action );
							dst.setType( type );
							dst.putExtra( Intent.EXTRA_STREAM, uri );
							return dst;
						}catch( Throwable ex ){
							log.trace( ex );
						}
					}
				}else if( Intent.ACTION_SEND_MULTIPLE.equals( action ) ){
					ArrayList< Uri > list_uri = src.getParcelableArrayListExtra( Intent.EXTRA_STREAM );
					if( list_uri == null ) return null;
					ArrayList< Uri > list_dst = new ArrayList<>();
					for( Uri uri : list_uri ){
						if( uri != null ){
							try{
								uri = saveToCache( uri );
								list_dst.add( uri );
							}catch( Throwable ex ){
								log.trace( ex );
							}
						}
					}
					if( list_dst.isEmpty() ) return null;
					Intent dst = new Intent( action );
					dst.setType( type );
					dst.putParcelableArrayListExtra( Intent.EXTRA_STREAM, list_dst );
					return dst;
				}
			}else if( Intent.ACTION_SEND.equals( action ) ){
				
				// Swarmアプリから送られたインテントは getType()==null だが EXTRA_TEXT は含まれている
				// EXTRA_TEXT の存在を確認してからtypeがnullもしくは text/plain なら受け取る
				
				String sv = src.getStringExtra( Intent.EXTRA_TEXT );
				if( ! TextUtils.isEmpty( sv ) && ( type == null || type.startsWith( "text/" ) ) ){
					String subject = src.getStringExtra( Intent.EXTRA_SUBJECT );
					if( ! TextUtils.isEmpty( subject ) ){
						sv = subject + " " + sv;
					}
					Intent dst = new Intent( action );
					dst.setType( "text/plain" );
					dst.putExtra( Intent.EXTRA_TEXT, sv );
					return dst;
				}
			}
		}catch( Throwable ex ){
			log.trace( ex );
		}
		return null;
	}
	
	private Uri saveToCache( Uri uri ) throws Throwable{
		
		// prepare cache directory
		File cache_dir = getCacheDir();
		//noinspection ResultOfMethodCallIgnored
		cache_dir.mkdirs();
		
		String name = "img." + Long.toString( System.currentTimeMillis() ) + "." + Utils.digestSHA256( uri.toString() );
		
		File dst = new File( cache_dir, name );
		
		FileOutputStream os = new FileOutputStream( dst );
		try{
			InputStream is = getContentResolver().openInputStream( uri );
			if( is == null )
				throw new RuntimeException( "getContentResolver.openInputStream returns null." );
			try{
				IOUtils.copy( is, os );
			}finally{
				IOUtils.closeQuietly( is );
			}
		}finally{
			IOUtils.closeQuietly( os );
		}
		return FileProvider.getUriForFile( this, App1.FILE_PROVIDER_AUTHORITY, dst );
	}
	
	private void sweepOldCache(){
		// sweep old cache
		try{
			// prepare cache directory
			File cache_dir = getCacheDir();
			//noinspection ResultOfMethodCallIgnored
			cache_dir.mkdirs();
			
			long now = System.currentTimeMillis();
			for( File f : cache_dir.listFiles() ){
				try{
					if( f.isFile()
						&& f.getName().startsWith( "img." )
						&& now - f.lastModified() >= 86400000L
						){
						//noinspection ResultOfMethodCallIgnored
						f.delete();
					}
				}catch( Throwable ex ){
					log.trace( ex );
				}
			}
		}catch( Throwable ex ){
			log.trace( ex );
		}
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/ActCallback.kt
package jp.juggler.subwaytooter

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import androidx.core.splashscreen.SplashScreen.Companion.installSplashScreen
import jp.juggler.subwaytooter.pref.FILE_PROVIDER_AUTHORITY
import jp.juggler.util.*
import jp.juggler.util.data.digestSHA256Hex
import jp.juggler.util.log.LogCategory
import jp.juggler.util.log.showToast
import java.io.File
import java.io.FileOutputStream
import java.util.*
import java.util.concurrent.atomic.AtomicReference

class ActCallback : AppCompatActivity() {

    companion object {
        private val log = LogCategory("ActCallback")

        internal val lastUri = AtomicReference<Uri>(null)
        internal val sharedIntent = AtomicReference<Intent>(null)

        private fun String?.isMediaMimeType() = when {
            this == null -> false
            this.startsWith("image/") -> true
            this.startsWith("video/") -> true
            this.startsWith("audio/") -> true
            else -> false
        }

        @Volatile
        private var uriFromApp: Uri? = null

        fun setUriFromApp(uri: Uri?) {
            synchronized(this) {
                uriFromApp = uri
            }
        }

        fun containsUriFromApp(uri: Uri?) =
            synchronized(this) {
                uri == uriFromApp
            }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        installSplashScreen()
        App1.setActivityTheme(this)

        var intent = this.intent
        log.d("onCreate flags=0x${intent?.flags?.toString(radix = 16)}")
        super.onCreate(savedInstanceState)

        when {
            intent == null -> {
                // 多分起きないと思う
            }

            (intent.flags and Intent.FLAG_ACTIVITY_LAUNCHED_FROM_HISTORY) != 0 -> {
                // 履歴から開いた場合はIntentの中味を読まない
            }

            else -> {
                val action = intent.action
                val type = intent.type
                // ACTION_SEND か ACTION_SEND_MULTIPLE
                // ACTION_VIEW かつ  type が 画像かビデオか音声
                when {
                    Intent.ACTION_SEND == action ||
                            Intent.ACTION_SEND_MULTIPLE == action ||
                            (Intent.ACTION_VIEW == action && type.isMediaMimeType()) -> {

                        // Google Photo などから送られるIntentに含まれるuriの有効期間はActivityが閉じられるまで
                        // http://qiita.com/pside/items/a821e2fe9ae6b7c1a98c

                        // 有効期間を延長する
                        intent = remake(intent)
                        if (intent != null) {
                            sharedIntent.set(intent)
                        }
                    }

                    forbidUriFromApp(intent) -> {
                        // last_uriをクリアする
                        lastUri.set(null)
                        // ダイアログを閉じるまで画面遷移しない
                        return
                    }

                    else -> {
                        val uri = intent.data
                        if (uri != null) {
                            lastUri.set(uri)
                        }
                    }
                }
            }
        }

        // どうであれメイン画面に戻る
        afterDispatch()
    }

    private fun afterDispatch() {
        val intent = Intent(this, ActMain::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_LAUNCHED_FROM_HISTORY or Intent.FLAG_ACTIVITY_NEW_TASK)
        startActivity(intent)
        finish()
    }

    private fun copyExtraTexts(dst: Intent, src: Intent) {
        src.string(Intent.EXTRA_TEXT)
            ?.let { dst.putExtra(Intent.EXTRA_TEXT, it) }
        //
        src.string(Intent.EXTRA_SUBJECT)
            ?.let { dst.putExtra(Intent.EXTRA_SUBJECT, it) }
    }

    private fun remake(src: Intent): Intent? {

        sweepOldCache()

        try {
            val action = src.action
            val type = src.type

            if (type.isMediaMimeType()) {
                when (action) {
                    Intent.ACTION_VIEW -> {
                        src.data?.let { uriOriginal ->
                            try {
                                val uri = saveToCache(uriOriginal)
                                val dst = Intent(action)
                                dst.setDataAndType(uri, type)
                                copyExtraTexts(dst, src)
                                return dst
                            } catch (ex: Throwable) {
                                log.e(ex, "remake failed. src=$src")
                            }
                        }
                    }

                    Intent.ACTION_SEND -> {
                        var uri = src.getStreamUriExtra()
                            ?: return src // text/plainの場合
                        try {
                            uri = saveToCache(uri)

                            val dst = Intent(action)
                            dst.type = type
                            dst.putExtra(Intent.EXTRA_STREAM, uri)
                            copyExtraTexts(dst, src)
                            return dst
                        } catch (ex: Throwable) {
                            log.e(ex, "remake failed. src=$src")
                        }
                    }

                    Intent.ACTION_SEND_MULTIPLE -> {
                        val listUri = src.getStreamUriListExtra()
                            ?: return null
                        val listDst = ArrayList<Uri>()
                        for (uriOriginal in listUri) {
                            if (uriOriginal != null) {
                                try {
                                    val uri = saveToCache(uriOriginal)
                                    listDst.add(uri)
                                } catch (ex: Throwable) {
                                    log.e(ex, "remake failed. src=$src")
                                }
                            }
                        }
                        if (listDst.isEmpty()) return null
                        val dst = Intent(action)
                        dst.type = type
                        dst.putParcelableArrayListExtra(Intent.EXTRA_STREAM, listDst)
                        copyExtraTexts(dst, src)
                        return dst
                    }
                }
            } else if (Intent.ACTION_SEND == action) {

                // Swarmアプリから送られたインテントは getType()==null だが EXTRA_TEXT は含まれている
                // EXTRA_TEXT の存在を確認してからtypeがnullもしくは text/plain なら受け取る

                val sv = src.string(Intent.EXTRA_TEXT)
                if (sv?.isNotEmpty() == true && (type == null || type.startsWith("text/"))) {
                    val dst = Intent(action)
                    dst.type = "text/plain"
                    copyExtraTexts(dst, src)

                    return dst
                }
            }
        } catch (ex: Throwable) {
            log.e(ex, "remake failed. src=$src")
        }

        return null
    }

    @Throws(Throwable::class)
    private fun saveToCache(uri: Uri): Uri {

        // prepare cache directory
        val cacheDir = this.cacheDir

        cacheDir.mkdirs()

        val name =
            "img." + System.currentTimeMillis().toString() + "." + uri.toString().digestSHA256Hex()

        val dst = File(cacheDir, name)

        FileOutputStream(dst).use { outStream ->
            val source = contentResolver.openInputStream(uri)
                ?: error("getContentResolver.openInputStream returns null.")
            source.use { inStream ->
                inStream.copyTo(outStream)
            }
        }
        return FileProvider.getUriForFile(this, FILE_PROVIDER_AUTHORITY, dst)
    }

    private fun sweepOldCache() {
        // sweep old cache
        try {
            // prepare cache directory
            val cacheDir = this.cacheDir

            cacheDir.mkdirs()

            val now = System.currentTimeMillis()
            val files = cacheDir.listFiles()
            if (files != null) for (f in files) {
                try {
                    if (f.isFile &&
                        f.name.startsWith("img.") &&
                        now - f.lastModified() >= 86400000L
                    ) {
                        f.delete()
                    }
                } catch (ex: Throwable) {
                    log.e(ex, "sweepOldCache: delete item failed.")
                }
            }
        } catch (ex: Throwable) {
            log.e(ex, "sweepOldCache failed.")
        }
    }

    // return true if open app chooser dialog
    private fun forbidUriFromApp(src: Intent): Boolean {
        if (src.action != Intent.ACTION_VIEW) return false

        val uri = src.data ?: return false
        if (!containsUriFromApp(uri)) return false
        setUriFromApp(null)

        try {
            startActivity(Intent.createChooser(src, uri.toString()))
            finish()
        } catch (ex: Throwable) {
            showToast(ex, "can't open chooser for $uri")
        }
        return true
    }
}
